README.md
# PostBin

PostBin, a simple web service for testing and logging of the receival of WebHooks
(HTTP POST requests).

## Quick Start: Stand Alone

```shell
$ gem install postbin
$ postbin
== Starting PostBin on http://127.0.0.1:6969
== Sinatra/1.3.1 has taken the stage on 6969 for development with backup from Thin
>> Thin web server (v1.2.11 codename Bat-Shit Crazy)
>> Maximum connections set to 1024
>> Listening on 127.0.0.1:6969, CTRL+C to stop
```

When running postbin from the command line, requests are stored in a temporary
file database, they will be lost once the server terminates.

You can then submit a basic POST request using `curl` as follows:

```shell
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"foo":"bar", "hello":"world", "number": 42}' \
  http://127.0.0.1:6969/my/fake/path
```

Rar

## Quick Start: Rack Application

You can run a more permeant install by running PostBin as a Rack application:

```ruby
# example config.ru
require 'postbin'

# path of pstore file to use for storage.
pstore = File.expand_path(File.join(File.dirname(__FILE__), 'postbin.pstore'))

# start the server.
PostBin::Server.set :pstore_file, pstore
run PostBin::Server.new(pstore)
```

## Command Line Options

```text
Usage: postbin [options]

PostBin options:
  -v, --version            show version number
  -h, --help               show this message

Rack options:
  -s, --server SERVER      server (webrick, mongrel, thin, etc.)
  -a, --address HOST       listen on HOST address (default: 127.0.0.1)
  -p, --port PORT          use PORT number (default: 6969)
```

---

## Quick Start: Local Development

Local development/testing is handled by `rake` tasks for the most part:

```shell
# list available `rake` task
rake -T
# build and install gem
rake install
```

If you wish to build/test a set of uncommited changes:

```shell
# build gem
gem build postbin.gemspec
# install gem (adjust as needed)
gem install postbin-<VERSION>.gem
```

## Release Instructions

1. Bump the version number found in `lib/postbin/version.rb`.
2. Tag git repo with the same version number.
3. Publush `.gem` to https://rubygems.org/:

```shell
# ensure tests pass first
rake test
# build a new gem
rake build
# release the new gem
rake release
```


HISTORY.md
# History

This file contains the overall histroy/changelog of the project.

## HEAD

## 0.1.7 (2020-12-18)

* (#3 - @drnic - Dr Nic Williams) Add `thin` depndency to Gemspec.
* Documentation improvments and a some 'house keeping' commits to the repo.

## 0.1.6 (2012-01-04)

* Added support for Ruby 1.8.7
* Let rubygems deal with $LOAD_PATH
* Change JSON lib to `multi_json` gem.

## 0.1.0 (2011-11-22)

* Initial gem release, v0.1.0


test/post_test.rb
require 'test_helper'

class TestPost < MiniTest::Unit::TestCase

  def test_has_a_header_and_body
    post = PostBin::Post.new('header data', 'body data')
    assert_equal 'header data', post.headers, 'post headers not as we expect'
    assert_equal 'body data', post.body, 'post body not as we expect'
  end

  def test_has_received_at_timestamp
    post = PostBin::Post.new('headers', 'body')
    assert_kind_of Time, post.received_at
  end

  def test_we_can_set_received_at
    now = Time.now
    post = PostBin::Post.new('headers', 'body', now)
    assert_equal now, post.received_at
  end

  def test_double_equals_comparison_method
    now = Time.now
    post1 = PostBin::Post.new('headers', 'body', now)
    post2 = PostBin::Post.new('headers', 'body', now)
    assert_equal post1, post2, 'should be the same'
  end

end


test/server_test.rb
require 'test_helper'

class TestServer < MiniTest::Unit::TestCase
  def app
    PostBin::Server.new
    # use a new pstore file for each request.
    PostBin::Server.set :pstore_file, Tempfile.new(['postbin', 'pstore']).path
  end

  def test_roots_should_redirect_to_overview
    get '/'
    follow_redirect!
    assert_equal '/postbin/overview', last_request.path
    assert last_response.ok?

    get '/postbin'
    follow_redirect!
    assert_equal '/postbin/overview', last_request.path
    assert last_response.ok?
  end

  def test_overview_returns_200
    get '/postbin/overview'
    assert_equal 200, last_response.status, 'status code should be 200 OK'
  end

  def test_accepts_post_requests
    post '/hello', 'foobar'
    assert_equal 201, last_response.status, 'status code should be 201 Created'
  end

  def test_accepts_posts_to_long_urls
    post '/foo/bar/is/very/long.php', 'hello world'
    assert_equal 201, last_response.status, 'status code should be 201 Created'
  end

  def test_json_url_hits
    get '/postbin/hits'
    assert_equal 200, last_response.status, 'status code should be 200 OK'
    assert_includes last_response.headers['Content-Type'], 'application/json', 'Content-Type header should be json'
  end

  def test_json_url_posts
    get '/postbin/posts'
    assert_equal 200, last_response.status, 'status code should be 200 OK'
    assert_includes last_response.headers['Content-Type'], 'application/json', 'Content-Type header should be json'
  end

  def test_post_and_read_back
    # post some data.
    post '/cats', 'they miaow'
    get '/postbin/hits'
    assert_equal '{"/cats":1}', last_response.body

    # read it back.
    get '/postbin/posts/cats'
    post = MultiJson.decode(last_response.body)[0]
    assert_equal 'they miaow', post['body']
  end


end


test/test_helper.rb
#
# Test helper init script.
#

# --- load rubygems if needed ------------------------------------------------
require 'rubygems' unless defined?(Gem)

# --- load simplecov if the gem is installed ---------------------------------
begin
  require 'simplecov'

  # code coverage groups.
  SimpleCov.start do
    add_filter 'test/'
  end
rescue LoadError
  # dont run code coverage if gem isn't installed.
end

# --- load our dependencies using bundler ------------------------------------
require 'bundler/setup'
Bundler.setup
require 'minitest/autorun'
require 'minitest/pride'
require 'rack/test'

# --- load postbin lib -------------------------------------------------------
require 'postbin'

# --- extend main TestCase ---------------------------------------------------
class MiniTest::Unit::TestCase
  include Rack::Test::Methods
end


lib/postbin/storage.rb
module PostBin

  # Storage backend for PostBin, uses PStore under the hood.
  class Storage

    def initialize(pstore_file)
      # setup file database for storage.
      @db = PStore.new(pstore_file)
      @db.ultra_safe = true if @db.respond_to?(:ultra_safe)
    end

    # Store a post in the database.
    def store(url, post)
      @db.transaction do
        # init if no data exists.
        @db['urls'] ||= {}
        @db['urls'][url] ||= 0
        @db[url] ||= []
        # incr hit count.
        @db['urls'][url] += 1
        # store post.
        @db[url] << post
      end
      true
    end

    # Returns hash, key being url and value being number of posts received.
    def hits
      @db.transaction(true) do
        @db['urls'] || {}
      end
    end

    # Returns array of urls that have been posted to.
    def urls
      @db.transaction(true) do
        (@db['urls'] || {}).keys
      end
    end

    # Returns array of posts for the given url.
    def posts(url)
      @db.transaction(true) do
        @db[url] || []
      end
    end

  end
end


lib/postbin/server.rb
module PostBin
  class Server < Sinatra::Base

    dir = File.dirname(File.expand_path(__FILE__))

    set :views,  "#{dir}/server/views"
    set :public_folder, "#{dir}/server/public"
    set :static, true

    # by default we store in a temp file.
    set :pstore_file, Tempfile.new(['postbin', 'pstore']).path

    # Redirect lost users.
    get '/?' do;         redirect '/postbin/overview'; end
    get '/postbin/?' do; redirect '/postbin/overview'; end

    # Display main overview.
    get '/postbin/overview' do
      storage = PostBin::Storage.new(settings.pstore_file)
      # pull out all urls and there posts.
      @url_hits = storage.hits

      @url_posts = @url_hits.inject({}) do |result, data|
        url, hits = *data
        result[url] = storage.posts(url)
        result
      end

      erb :overview
    end

    # Display urls and there hit count as JSON.
    get '/postbin/hits' do
      storage = PostBin::Storage.new(settings.pstore_file)

      content_type :json
      MultiJson.encode(storage.hits)
    end

    # Display posts for the given URL as JSON.
    get %r{/postbin/posts(.*)} do
      url = params[:captures].first
      storage = PostBin::Storage.new(settings.pstore_file)

      content_type :json
      MultiJson.encode(storage.posts(url))
    end

    # Catch all for post requests.
    post '*' do
      storage = PostBin::Storage.new(settings.pstore_file)
      post = PostBin::Post.new(client_request_headers, request.body.read)
      storage.store(request.path, post)
      status 201
    end

    # Runs a server on local machine, called by command line executable.
    def self.run_command_line!(argv)
      options = parse_args(argv)
      $stderr.puts "== Starting PostBin on http://#{options[:bind]}:#{options[:port]}"
      run!(options)
    end

    # Parse command line args.
    def self.parse_args(argv)
      # default options.
      options = { :bind => '127.0.0.1', :port => 6969, :server => 'thin', :enviroment => :production }

      # available options.
      opts = OptionParser.new('', 24, '  ') do |opts|
        opts.banner = 'Usage: postbin [options]'
        opts.separator ''
        opts.separator 'PostBin options:'
        opts.on('-v', '--version', 'show version number') { $stderr.puts 'PostBin ' + PostBin::Version; exit }
        opts.on('-h', '--help', 'show this message') { $stderr.puts opts; exit; }
        opts.separator ''
        opts.separator 'Rack options:'
        opts.on('-s', '--server SERVER', 'server (webrick, mongrel, thin, etc.)') { |s| options[:server] = s }
        opts.on('-a', '--address HOST', 'listen on HOST address (default: 127.0.0.1)') { |host| options[:bind] = host }
        opts.on('-p', '--port PORT', 'use PORT number (default: 6969)') { |port| options[:port] = port }
      end.parse!(argv)

      options
    end

    private
    # Returns an array of all client side HTTP request headers.
    def client_request_headers
      # POST /some/url HTTP/1.1
      # Accept: application/json
      # Content-Type: application/json
      headers = request.env.select { |k,v| k.start_with? 'HTTP_' } \
        .collect { |pair| [ pair[0].sub(/^HTTP_/, ''), pair[1] ] } \
        .collect { |pair| pair.join(': ') }

      headers
    end

  end
end
